hwcomposer/gralloc_handler.cpp

```diff
@@ -0,0 +1,334 @@
+        format = display->target_layer_handle_ext.format;
+        pixel_stride = display->target_layer_handle_ext.stride;
+        width = display->target_layer_handle_ext.width;
+        height = display->target_layer_handle_ext.height;
+    } else {
+        format = display->layer_handles_ext[pos].format;
+        pixel_stride = display->layer_handles_ext[pos].stride;
+        width = display->layer_handles_ext[pos].width;
+        height = display->layer_handles_ext[pos].height;
+    }
+
+    if (!width)
+        width = layer->displayFrame.right - layer->displayFrame.left;
+    if (!height)
+        height = layer->displayFrame.bottom - layer->displayFrame.top;
+
+    return {height, width, pixel_stride, format};
+}
+buffer_metadata get_buffer_metadata_gbm(display *, hwc_layer_1_t *layer, size_t) {
+    auto handle = reinterpret_cast<const gralloc_handle_t *>(layer->handle);
+    return {
+            handle->height,
+            handle->width,
+            handle->stride,
+            handle->format
+    };
+}
+buffer_metadata get_buffer_metadata_cros(display *, hwc_layer_1_t *layer, size_t) {
+    auto handle = reinterpret_cast<const cros_gralloc_handle *>(layer->handle);
+    return {
+            handle->height,
+            handle->width,
+            handle->pixel_stride,
+            static_cast<uint32_t>(handle->droid_format)
+    };
+}
+
+std::unique_ptr<buffer> create_buffer_generic(display *display, const buffer_metadata& metadata, buffer_handle_t handle) {
+    return create_shm_wl_buffer(display, metadata, handle);
+}
+std::unique_ptr<buffer> create_buffer_gbm(display *display, const buffer_metadata& metadata, buffer_handle_t handle) {
+    auto *drm_handle = reinterpret_cast<const gralloc_handle_t *>(handle);
+    return create_dmabuf_wl_buffer(display, metadata, drm_handle->prime_fd, -1 /* compute drm format */, drm_handle->stride, 0 /* offset */, drm_handle->modifier, handle);
+}
+std::unique_ptr<buffer> create_buffer_cros(display *display, const buffer_metadata& metadata, buffer_handle_t handle) {
+    auto *cros_handle = reinterpret_cast<const cros_gralloc_handle *>(handle);
+    return create_dmabuf_wl_buffer(display, metadata, cros_handle->fds[0], cros_handle->format, cros_handle->strides[0], cros_handle->offsets[0], cros_handle->format_modifier, handle);
+}
+std::unique_ptr<buffer> create_buffer_android(display *display, const buffer_metadata& metadata, buffer_handle_t handle) {
+    return create_android_wl_buffer(display, metadata, handle);
+}
+
+void update_shm_buffer_generic(display *display, buffer *buffer) {
+    display->egl_work_queue.emplace_back(std::bind(egl_render_to_pixels, display, buffer));
+    sem_post(&display->egl_go);
+    sem_wait(&display->egl_done);
+}
+void update_shm_buffer_default(display *, buffer *buffer) {
+    void *data;
+    uint32_t shm_stride, src_stride;
+    android::Rect bounds(buffer->metadata.width, buffer->metadata.height);
```

Comment:
`#include <ui/Rect.h>` explicitely in case it's not recursively included (it's not in older AOSP, I was testing on lineage-16.0 now)